python_modules/dagster/dagster/_config/field_utils.py

```diff
@@ -459,20 +459,31 @@ def _convert_potential_field(
     return Field(_convert_potential_type(original_root, potential_field, stack))
 
 
+def _config_dictionary_from_values_inner(obj: Any):
+    from dagster._config.structured_config import Config
+
+    if isinstance(obj, dict):
+        return {k: _config_dictionary_from_values_inner(v) for k, v in obj.items() if v is not None}
+    elif isinstance(obj, list):
+        return [_config_dictionary_from_values_inner(v) for v in obj]
+    elif isinstance(obj, EnvVar):
+        return {"env": str(obj)}
+    elif isinstance(obj, Config):
+        return {
+            k: _config_dictionary_from_values_inner(v)
+            for k, v in obj.dict().items()
+            if not k.startswith("_")
+        }
```

Comment:
also to be clear this change is sort of separate from the title (refactoring resources to be immutable), as it expands support of the coercion to include lists and strongly typed config objects, correct?

Reply:
Yes, exactly. Added a comment to this effect, and updated PR description just to make it more clear.